src/algebra/finite_map.rs

```diff
@@ -317,35 +315,109 @@ where
     }
 }
 
+impl<KeyType, ValueType> Add<FiniteHashMap<KeyType, ValueType>>
+    for FiniteHashMap<KeyType, ValueType>
+where
+    KeyType: KeyProperties,
+    ValueType: GroupValue,
+{
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        fn do_add<KeyType, ValueType>(
+            mut this: FiniteHashMap<KeyType, ValueType>,
+            other: FiniteHashMap<KeyType, ValueType>,
+        ) -> FiniteHashMap<KeyType, ValueType>
+        where
+            KeyType: KeyProperties,
+            ValueType: GroupValue,
+        {
+            for (k, v) in other.value {
+                let entry = this.value.entry(k);
+                match entry {
+                    Entry::Vacant(e) => {
+                        e.insert(v);
+                    }
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().add_assign(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        }
+                    }
+                }
+            }
+            this
+        }
+        if self.support_size() > other.support_size() {
+            do_add(self, other)
+        } else {
+            do_add(other, self)
+        }
+    }
+}
 impl<KeyType, ValueType> AddByRef for FiniteHashMap<KeyType, ValueType>
 where
     KeyType: KeyProperties,
     ValueType: GroupValue,
 {
     fn add_by_ref(&self, other: &Self) -> Self {
-        let mut result = self.clone();
-        for (k, v) in &other.value {
-            // TODO: unfortunately there is no way to avoid this k.clone() currently.
-            // See also note on 'insert' below.
-            let entry = result.value.entry(k.clone());
+        fn do_add<KeyType, ValueType>(
+            mut this: FiniteHashMap<KeyType, ValueType>,
+            other: &FiniteHashMap<KeyType, ValueType>,
+        ) -> FiniteHashMap<KeyType, ValueType>
+        where
+            KeyType: KeyProperties,
+            ValueType: GroupValue,
+        {
+            for (k, v) in &other.value {
+                // TODO: unfortunately there is no way to avoid this k.clone() currently.
+                // See also note on 'insert' below.
+                let entry = this.value.entry(k.clone());
+                match entry {
+                    Entry::Vacant(e) => {
+                        e.insert(v.clone());
+                    }
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().add_assign_by_ref(v);
+                        if e.get().is_zero() {
+                            e.remove_entry();
+                        }
+                    }
+                }
+            }
+            this
+        }
+
+        if self.support_size() > other.support_size() {
+            do_add(self.clone(), other)
```

Comment:
both these do the same amount of work algorithmically. this optimization is justified only if clone is much faster than inserting all entries.

Reply:
That's right, cloning the map should be more efficient than element-by-element insertion.